fms/training/plugins.py

```diff
@@ -0,0 +1,228 @@
+        optimizer,
+        epoch: int,
+        metrics: Dict = {},
+        step: int = None,
+    ):
+        if "batch_size" in metrics and "input_length" in metrics:
+            self.tokens_seen += metrics["batch_size"] * metrics["input_length"]
+        if "loss" in metrics:
+            self.cum_loss += metrics["loss"]
+
+        current_time = datetime.now()
+        elapsed = (current_time - self.last_reported_time).total_seconds()
+
+        if step is None:
+            step = self.last_step + 1
+            self.last_step = 0
+            steps = step - self.last_reported_step
+            self.last_reported_step = 0
+        else:
+            self.last_step = step
+            if elapsed < self.seconds or step == self.last_reported_step:
+                return
+            steps = step - self.last_reported_step
+            self.last_reported_step = step
+
+        # TODO: aggregate these per-rank statistics when training with
+        # distributed. e.g.: dist.all_reduce(loss, op=dist.ReduceOp.SUM)
+        # for loss, tokens_seen, etc.
+        to_report = {}
+        if "loss" in metrics:
+            to_report["loss"] = f"{metrics['loss']:.4f}"
+        to_report["avg_loss"] = f"{self.cum_loss / steps:.4f}"
+
+        more_metrics = {
+            "tok/stp": f"{self.tokens_seen / steps:,.1f}",
+            "tok/s": f"{self.tokens_seen / elapsed:,.1f}",
+        }
+        if torch.cuda.is_available() and utils.has_package("pynvml"):
+            nvidia_metrics = {
+                "gpu_mem_use": f"{torch.cuda.memory_usage()}%",
+                "gpu_utzn": f"{torch.cuda.utilization()}%",
+            }
+            more_metrics.update(nvidia_metrics)
+
+        self.last_reported_time = current_time
+        self.tokens_seen = 0
+        self.cum_loss = 0
+        to_report.update(more_metrics)
+        self.writer(epoch, step, current_time, to_report)
+
+
+class Checkpointer(TrainerPlugin):
+    """
+    A training plugin to write checkpoints.
+    TODO: This will require changes to handle distributed checkpoints.
+    """
+
+    def __init__(
+        self,
+        save_dir: Path = Path("./checkpoints"),
+        steps: int = None,
+        name: str = None,
+        group: dist.ProcessGroup = None,
+    ):
+        super().__init__(steps)
+        os.makedirs(save_dir, exist_ok=True)
+        if type(save_dir) == str:
+            save_dir = os.path.expanduser(save_dir)
+            save_dir = Path(save_dir)
+        self.save_dir = save_dir
+        self.group = group
+        self.name = name
+
+    # TODO: this probably also needs to accept a dataset since we want to
+    # support checkpointable datasets.
+    # TODO: end of epoch checkpoints should probably consolidate to one
+    # rank when using FSDP.
+    def step(
+        self,
+        model: nn.Module,
+        optimizer,
+        epoch: int,
+        metrics: Dict = {},
+        step: int = None,
+    ):
+        if not self.run(step):
```

Comment:
should this somehow be getting called implicitly? I would expect that this function would be running outside somewhere and if returning true, would call this step function

Reply:
I was initially intending to have it be implicit, but for plugins like "metricreporter" we want to record information on every step and only skip after the information is recorded. See the implementation of `MetricReporter` for more detail on why making this implicit would cause issues.